src/lsm/block_free_set.zig

```diff
@@ -1,37 +1,421 @@
 const std = @import("std");
 const assert = std.debug.assert;
 const mem = std.mem;
+const DynamicBitSetUnmanaged = std.bit_set.DynamicBitSetUnmanaged;
+const MaskInt = DynamicBitSetUnmanaged.MaskInt;
+const config = @import("../config.zig");
+const RingBuffer = @import("../ring_buffer.zig").RingBuffer(
+    usize,
+    config.cache_line_size / @sizeOf(usize), // 1 cache line of recently-freed blocks
+    .array,
+);
+const EWAH = @import("../ewah.zig").EWAH(usize);
 
-/// The 0 address is reserved for usage as a sentinel and will never be returned
-/// by acquire().
-/// TODO add encode/decode function for run-length encoding to/from SuperBlock.
+/// The 0 address is reserved for usage as a sentinel and will never be returned by acquire().
+///
+/// Set bits indicate free blocks, unset bits are allocated.
 pub const BlockFreeSet = struct {
-    /// Bits set indicate free blocks
-    free: std.bit_set.DynamicBitSetUnmanaged,
+    // Each bit of `index` is the OR of `shard_size` bits of `blocks`.
+    // That is, if a shard has any free blocks, the corresponding index bit is set.
+    index: DynamicBitSetUnmanaged,
+    blocks: DynamicBitSetUnmanaged,
+    // A fast cache of the 0-indexed bits (not 1-indexed addresses) of recently freed blocks.
+    recent: RingBuffer,
+
+    // Fixing the shard size to a constant rather than varying the shard size (but
+    // guaranteeing the index always a multiple of 64B) means that the top-level index
+    // may have some unused bits. But the shards themselves are always a multiple of
+    // the word size. In practice the tail end of the index will be accessed less
+    // frequently than the head/middle anyway.
+    //
+    // Each shard is 8 cache lines because the CPU line fill buffer can fetch 10 lines in parallel. And 8 is fast for division when computing the shard of a block.
+    // Since the shard is scanned sequentially, the prefetching amortizes the cost of the single
+    // cache miss. It also reduces the size of the index.
+    //
+    // e.g. 10TiB disk ÷ 64KiB/block ÷ 512*8 blocks/shard ÷ 8 shards/byte = 5120B index
+    const shard_cache_lines = 8;
+    const shard_size = shard_cache_lines * config.cache_line_size * @bitSizeOf(u8);
+    comptime {
+        assert(shard_size == 4096);
+        assert(@bitSizeOf(MaskInt) == 64);
+    }
+
+    pub fn init(allocator: *mem.Allocator, blocks_count: usize) !BlockFreeSet {
+        assert(shard_size <= blocks_count);
+        assert(blocks_count % shard_size == 0);
+        assert(blocks_count % @bitSizeOf(usize) == 0);
+
+        // Round up to ensure that every block bit is covered by the index.
+        const shards_count = div_ceil(blocks_count, shard_size);
```

Comment:
It would be great to restrict things around `blocks_count` so that we can use `@divExact()` here instead.

Reply:
It already was safe for `@divExact`, I just forgot to change it :stuck_out_tongue_closed_eyes: